packages/modals/src/elements/TooltipModal/TooltipModal.spec.tsx

```diff
@@ -0,0 +1,185 @@
+    });
+
+    it('renders RTL placement correctly', async () => {
+      const { container, getByTestId, rerender } = renderRtl(<div />);
+
+      await act(async () => {
+        await rerender(
+          <TooltipModal referenceElement={container} placement="start" data-test-id="modal">
+            <TooltipModal.Title>Title</TooltipModal.Title>
+            <TooltipModal.Body>Content</TooltipModal.Body>
+            <TooltipModal.Footer>
+              <TooltipModal.FooterItem>
+                <button>Finish</button>
+              </TooltipModal.FooterItem>
+            </TooltipModal.Footer>
+            <TooltipModal.Close aria-label="Close" />
+          </TooltipModal>
+        );
+      });
+
+      expect(getByTestId('modal').parentElement).toHaveAttribute('data-popper-placement', 'right');
+    });
+  });
+
+  it('passes ref to underlying DOM element', async () => {
+    const ref = React.createRef<HTMLDivElement>();
+    const { container, getByTestId, rerender } = render(<div />);
+
+    await act(async () => {
+      await rerender(
+        <TooltipModal referenceElement={container} ref={ref} data-test-id="tooltip-modal" />
+      );
+    });
+
+    expect(getByTestId('tooltip-modal')).toBe(ref.current);
+  });
+
+  it('applies backdropProps to Backdrop element', async () => {
+    const { container, getByTestId, rerender } = render(<div />);
+
+    await act(async () => {
+      await rerender(
+        <TooltipModal referenceElement={container} backdropProps={{ 'data-test-id': 'backdrop' }} />
+      );
+    });
+
+    expect(getByTestId('backdrop')).not.toBeNull();
+  });
+
+  it('applies title a11y attributes to Title element', async () => {
+    const { container, getByText, rerender } = render(<div />);
+
+    await act(async () => {
+      await rerender(
+        <TooltipModal referenceElement={container} id={TOOLTIP_MODAL_ID}>
+          <TooltipModal.Title>title</TooltipModal.Title>
+        </TooltipModal>
+      );
+    });
+
+    expect(getByText('title')).toHaveAttribute('id', `${TOOLTIP_MODAL_ID}--title`);
+  });
+
+  it('applies content a11y attributes to Body element', async () => {
+    const { container, getByText, rerender } = render(<div />);
+
+    await act(async () => {
+      await rerender(
+        <TooltipModal referenceElement={container} id={TOOLTIP_MODAL_ID}>
+          <TooltipModal.Body>body</TooltipModal.Body>
+        </TooltipModal>
+      );
+    });
+
+    expect(getByText('body')).toHaveAttribute('id', `${TOOLTIP_MODAL_ID}--content`);
+  });
+
+  it('applies close a11y attributes to Close element', async () => {
+    const { container, getByRole, rerender } = render(<div />);
+
+    await act(async () => {
+      await rerender(
+        <TooltipModal referenceElement={container}>
+          <TooltipModal.Close />
+        </TooltipModal>
+      );
+    });
+
+    expect(getByRole('button')).toHaveAttribute('aria-label', 'Close modal');
+  });
+
+  describe('onClose()', () => {
```

Comment:
I think asserting against function calls and using [`rerender`](https://testing-library.com/docs/react-testing-library/api#rerender) goes against Testing Library's [guiding principles](https://testing-library.com/docs/guiding-principles) as their guiding princples favor tests "that closely resemble how your web pages are used". 

From what I understand of the guiding principles, asserting against function calls and re-renders rank lower in terms of importance when compared to asserting against UI elements and simulating user events. 

What do you think about rewriting these tests to more align with the Testing Library guiding principles? Here's how I think a test that aligns with the guiding principles might look like. It is a test for asserting that the tooltip modal closes when clicking the close element.

```tsx
describe('Tooltip Modal', () => {
  const BasicTooltipModal = () => {
    const [isOpen, setIsOpen] = React.useState(false);
    const buttonRef = React.useRef<HTMLButtonElement>(null);

    return (
      <>
        <button ref={buttonRef} onClick={() => setIsOpen(!isOpen)}>
          Open Tooltip Modal
        </button>
        <TooltipModal
          hasArrow={true}
          placement="bottom-start"
          onClose={() => setIsOpen(false)}
          referenceElement={isOpen && buttonRef.current ? buttonRef.current : undefined}
        >
          <TooltipModal.Title>Header content</TooltipModal.Title>
          <TooltipModal.Body>Body content</TooltipModal.Body>
          <TooltipModal.Footer>Footer content</TooltipModal.Footer>
          <TooltipModal.Close aria-label="Close" />
        </TooltipModal>
      </>
    );
  };

  it('can close the tooltip modal by clicking the close element', async () => {
    const { getByText, queryByText, getByLabelText } = render(<BasicTooltipModal />);
    const button = getByText('Open Tooltip Modal');

    // Given: a tooltip modal is not rendered
    expect(queryByText('Header content')).toBeNull();
    expect(queryByText('Body content')).toBeNull();
    expect(queryByText('Footer content')).toBeNull();

    // When: the user clicks a button to open the tooltip modal
    await act(async () => {
      userEvent.click(button);
    });

    // Then: the tooltip modal is rendered in the document
    expect(getByText('Header content')).toBeInTheDocument();
    expect(getByText('Body content')).toBeInTheDocument();
    expect(getByText('Footer content')).toBeInTheDocument();

    // And when: the user closes the tooltip modal
    const close = getByLabelText('Close');
    userEvent.click(close);

    // Then: the tooltip modal is not rendered
    expect(queryByText('Header content')).toBeNull();
    expect(queryByText('Body content')).toBeNull();
    expect(queryByText('Footer content')).toBeNull();
  });
});
```

The test above could be written in different flavors. For example, with different matchers. Or can be abstracted differently. I think the main goal is that it aligns with the guiding principles by asserting against UI elements (not function calls) and simulating events (not manually calling `rerender`).

Reply:
The technical reason for this testing structure is to allow `forwardRef` and `usePopper` to accomplish state updates properly. The implementation above produces failures and warnings similar to:

> Warning: An update to ForwardRef inside a test was not wrapped in act(...)

https://github.com/popperjs/react-popper/issues/368 includes some additional info

Reply:
I ran the test implementation above and it didn't produce failures or warnings. The test passes and makes the correct assertions before and after each user event simulation ✅.

However, I noticed that there is a warning (not error) regarding an update to `ForwardRef` when the tooltip modal implementation broke. But the warning is caused by a bug in `react-popper` and is only visible when the Tooltip Modal implementation is broken. I don't think that warrants the test to be structured differently than what the Testing Library guiding principles suggest. Also, as long as the test is passing, the warning won't be encountered.

Alternatively, I think the test can be kept as-is with a code comment pointing to popperjs/react-popper#368 so that the tests in this file can be rewritten when `react-popper` fixes the associated bug.

Reply:
Leaving a note for posterity... from our offline chat I needed to change the order of some `await/async` commands for this to pass correctly. Tests are now updated.